0.0.81.argocd/pkg/argocd/api.go

```diff
@@ -0,0 +1,271 @@
+	argoCDContainerName  = "application-controller"
+
+	argoClientSecretName = "argocd/client_secret" //nolint:gosec
+	argoSecretKeyName    = "argocd/secret_key"    //nolint:gosec
+	argoSecretName       = "argocd-secret"
+	argoPrivateKeyName   = "argocd-privatekey"
+)
+
+var (
+	expectedContainerImagePreUpgradeVersion = semver.MustParse("1.7.2")
+	targetContainerImageVersion             = semver.MustParse("2.1.7")
+)
+
+// ArgoCD is a sample okctl component
+type ArgoCD struct {
+	log     logger.Logger
+	dryRun  bool
+	confirm bool
+	kubectl Kubectl
+}
+
+var errNothingToDo = errors.New("nothing to do")
+
+// Upgrade upgrades the component
+func (a ArgoCD) Upgrade() error {
+	a.log.Info("Upgrading ArgoCD")
+
+	err := a.preflight()
+	if err != nil {
+		if errors.Is(err, errNothingToDo) {
+			return nil
+		}
+
+		return fmt.Errorf("running preflight checks: %w", err)
+	}
+
+	o, err := initializeOkctl()
+	if err != nil {
+		return fmt.Errorf("initializing: %w", err)
+	}
+
+	clusterID := getClusterID(o)
+	state := o.StateHandlers(o.StateNodes())
+
+	services, err := o.ClientServices(state)
+	if err != nil {
+		return err
+	}
+
+	err = a.partiallyRemoveArgoCD(services, clusterID)
+	if err != nil {
+		return fmt.Errorf("removing ArgoCD: %w", err)
+	}
+
+	err = a.createArgoCD(services, clusterID, state, o.Declaration)
+	if err != nil {
+		return fmt.Errorf("creating ArgoCD: %w", err)
+	}
+
+	err = a.postflight()
+	if err != nil {
+		return fmt.Errorf("running postflight checks: %w", err)
+	}
+
+	a.log.Info("Upgrading ArgoCD done!")
+
+	return nil
+}
+
+func (a ArgoCD) preflight() error {
+	isInstalled, err := a.kubectl.hasDeployment(argoCDNamespace, argoCDDeploymentName)
+	if err != nil {
+		return fmt.Errorf("getting argocd: %w", err)
+	}
+
+	if !isInstalled {
+		a.log.Info("ArgoCD is not installed, not doing anything")
+		return nil
+	}
+
+	currentVersion, err := a.kubectl.getDeploymentImageVersion(argoCDNamespace, argoCDDeploymentName, argoCDContainerName)
+	if err != nil {
+		return fmt.Errorf("getting current image version: %w", err)
+	}
+
+	err = a.validateVersion(expectedContainerImagePreUpgradeVersion, currentVersion)
+	if err != nil {
+		if currentVersion.GreaterThan(targetContainerImageVersion) || currentVersion.Equal(targetContainerImageVersion) {
+			a.log.Infof("Current version is %s, ignoring upgrade\n", currentVersion.String())
+
+			return errNothingToDo
+		}
+
+		return fmt.Errorf("unexpected version installed: %w", err)
+	}
+
+	return nil
+}
+
+func (a ArgoCD) validateVersion(expected *semver.Version, actual *semver.Version) error {
+	if !actual.Equal(expected) {
+		return fmt.Errorf("expected %s, got %s", expected, actual)
+	}
+
+	return nil
+}
+
+func (a ArgoCD) partiallyRemoveArgoCD(services *core.Services, clusterID api.ID) error {
+	if a.dryRun {
+		return nil
+	}
+
+	a.log.Info("Removing Old ArgoCD")
+
+	a.log.Debug("Deleting Helm Release")
+	err := services.Helm.DeleteHelmRelease(context.Background(), client.DeleteHelmReleaseOpts{
```

Comment:
By removing ArgoCD first, the user will lose their applied argocd-application.yamls